File: 0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        if(n == 0) return 0;
        Set<Character> set = new HashSet<>();
        int size = 0;
        int left = 0;
        int right = 0;
        for(char a : s.toCharArray()){
            while(set.contains(a)){
                set.remove(s.charAt(left++));
            }
            set.add(a);
            right++;
            size = Math.max(size, right - left);
        }
        return size;
    }
}

File: 0003-longest-substring-without-repeating-characters/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let n = s.length;
    if(n === 0) return 0;
    let set = new Set();
    let size = 0;
    let left = 0;
    let right = 0;

    for(let char of s){
        while(set.has(char)){
            set.delete(s[left++]);
        }
        set.add(char);
        right++;
        size = Math.max(size, right - left);
    }
    return size;
};
